src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -673,13 +677,35 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
 
             // Read the first vector unaligned.
 
-            firstVector = Sse2.LoadVector128((ushort*)pBuffer); // unaligned load
+            if (Sse2.IsSupported)
+            {
+                firstVector = Sse2.LoadVector128((ushort*)pBuffer); // unaligned load
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                firstVector = AdvSimd.LoadVector128((ushort*)pBuffer); // unaligned load
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
 
             // The operation below forces the 0x8000 bit of each WORD to be set iff the WORD element
-            // has value >= 0x0800 (non-ASCII). Then we'll treat the vector as a BYTE vector in order
+            // has value >= 0x0080 (non-ASCII). Then we'll treat the vector as a BYTE vector in order
             // to extract the mask. Reminder: the 0x0080 bit of each WORD should be ignored.
 
-            currentMask = (uint)Sse2.MoveMask(Sse2.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte());
+            if (Sse2.IsSupported)
+            {
+                currentMask = (uint)Sse2.MoveMask(Sse2.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte());
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                currentMask = Unicode.Utf16Utility.GetNonAsciiBytes(AdvSimd.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte(), bitmask);
```

Comment:
You should check out excellent feedback from @TamarChristinaArm  in https://github.com/dotnet/runtime/pull/39050#issuecomment-662458778 and https://github.com/dotnet/runtime/pull/39050#issuecomment-668699538 about optimizing this code. I think you should follow the same because in the end, all you are doing this calling `BitOperations.TrailingZeroCount()` on it.

Reply:
Agreed with splitting it off. In this case it looks like this code only cares about the index of the first non-zero element. So we can do even better here.

If you use a mask `0x0f00` and `BIC` you can get a much more efficient sequence (see https://github.com/ARM-software/optimized-routines/blob/224cb5f67b71757b99fe1e10b5a437c17a1d733c/string/aarch64/strlen.S#L164)

essentially

```
sshr	v1.16b, v1.16b, 7
bic	v1.8h, 0x0f, lsl 8
umaxp	v1.16b, v1.16b, v1.16b
fmov	x1, d1
clz	x0, x1
```

as the sequence to get the first element that has the `msb` set. Of course for the cases where it's just doing `if (currentMask != 0)` you just need a `maxp` and `fmov` as I explained in the other post.

Reply:
this is for little endian btw, for big-endian you need a slight variantion. though this can be avoided loading with `LD1` instead of `LDR` (which I think is what `LoadVector128` does here).